### PR TITLE
Optimizations to column sort

### PR DESCRIPTION
**Pull request by peeyushgupta**
Merged step 2(transpose) with 1(sort) and step 6(shift) with step 5(sort).
The merging of steps reduces time since it extends the use of the pinned
column table.

### CODE REVIEW

**Comment by Danappelxx** [comment]
I get a segfault when I run tests without the `PIN_TABLE` flag.

Works fine with the `PIN_TABLE` flag enabled, I get about a 5% performance improvement (tested on just one run).

File: enclave/column_sort.cpp [hunk 1]
```diff
@@ -510,119 +512,92 @@ int column_sort_table_parallel(data_base_t *db, table_t *table, int column, int
...
-
+		/* Transpose s[i] column table into s transposed tables  */
 		for (unsigned int j = 0; j < r; j ++) {
-			unsigned long seq; 
+			unsigned long seq;
 
 			// Read old row
-			ret = read_row(s_tables[i], j, row);
+
+			ret = get_pinned_row(s_tables[i], j, &b, &row_tmp);
```
Comment (peeyushgupta) [inline enclave/column_sort.cpp]: I missed to check `if (s_tables[i]->pinned_blocks)` condition here

File: enclave/column_sort.cpp [hunk 2]
```diff
@@ -770,6 +747,36 @@ int column_sort_table_parallel(data_base_t *db, table_t *table, int column, int
 		ret = bitonic_sort_table_parallel(s_tables[i], column, tid, num_threads);
 #endif
 		barrier_wait(&column_barrier, &column_lsense, tid, num_threads);
+
+		for (unsigned int j = 0; j < r; j ++) {
+
+			/* Read row from s table */
+			ret = get_pinned_row(s_tables[i], j, &b, &row_tmp);
```
Comment (peeyushgupta) [inline enclave/column_sort.cpp]: I missed to check if (s_tables[i]->pinned_blocks) condition here

**Comment by peeyushgupta** [comment]
You are write the pull request has a bug the calls to `get_pinned_row` should be made when `PIN_TABLE` is enabled otherwise `read_row` should be used. Check my comments in the diff.

I am closing the PR as the improvement is not significant, but the code can be `probably` made simpler/smaller/faster by working along the lines of this PR.

File: enclave/column_sort.cpp (hunk 2 above)
Comment (Danappelxx) [reply enclave/column_sort.cpp]: Yep, I just caught that too.

**Comment by Danappelxx** [comment]
Sounds good, I'll see if unpinning tables without marking them dirty has a significant impact as well.

**Comment by peeyushgupta** [comment]
Yep, but it should be done in cases when the pinned table is never modified.

**Comment by Danappelxx** [comment]
Ok, unfortunately I don't think we can pin tables further than we currently do, and right now we only pin them for sorting so they have to be marked as dirty.

Potential for further pinning (no go's):
- For the first step, when the `s` column tables are built, the main table is too big to fit in memory so we cannot pin it.
- For transposing and un-transposing, we alternate between each `s` table so we would have to pin all of them, which again would not fit in memory.

**Comment by peeyushgupta** [comment]
Following is the set of tables we have (`1 big table, s small s_tables and s small st_tables`) 

First the big table is processed to create s small s_tables (in chunks of r rows) and when all s_table are created we pin them and sort them and unpin them one by one.
One optimization here can be that, we pin the s_table right after all the required rows of it are inserted
and sort it before inserting data into the next s_table, this should improve the time taken in pinning since most of the blocks of the s_table should already be in memory. However if you pin the 1st s_table after finishing  inserting in all the s_tables, the pinning of 1st s_table will require more times since there is less probable that the blocks of the 1st s_table are still available in memory.

Basically it'll merge the sorting part into the s_table creation part. Extending this further, if you see the transpose first uses 1st s_table completely and then moves on to the next s_table, therefore, we can also merge the transpose part to the previous steps. Now transpose can be done right after an s_table is sorted and is still pinned.

Again, these may or may not improve the time by much (as the PR was implementing some parts of the above mentioned method). You can probably finish some other tasks then, and come back to this one later or whatever Shantanu suggests.
